### ISSUE TITLE
Map: Megastudy row-specific vocabulary endpoint

### ISSUE THREAD

**Issue by d-callan** [VEuPathDB/EdaDataService#264]
The megastudy has variables, like species, which do not have a reliable vocabulary across the entire term. Instead the vocabulary might be study or collection specific, meaning each row of some downstream entity has a different expected vocabulary based on an upstream entity. We need a mechanism to find the expected vocabulary AFTER filters have been taken into account.

This could be a new metadata-style endpoint, which other plugins could call. It would need to know which ancestor variable contains the complete row-specific vocabulary, and what filters have been applied to the child variable (ex species).

**Comment by d-callan** [comment]
we may hard code the study entity as being the one we always look to to infer the vocabulary

**Comment by d-callan** [comment]
for more info, were planning to add a new annotation `hasStudyDependentVocabulary`. this will need incorporated into the reference metadata in edacommon presumably (dont think theres a separate ticket for that, not sure if we want one). the value of this annotation would be true/ false, and for vars where its true we need to be able to hit this endpoint which id imagine should:
1. accept a set of filters, and the var spec for the var in question
2. returns results in the entity space of the var in question, each row w a list of expected values/ vocabulary
3. ideally have the results meet the merge service api, in case we want to modify the data flow at some point to have the merge service automagically provide this for us. (disclaimer: idk how hard that is)
4. for each row in the output entity, use its upstream study entity to find all possible values for the var of interest in that study. then remove values explicitly filtered against.
5.

**Comment by dmgaldi** [comment]
I worry a little bit about merging the vocabulary rows into the entity space peformance-wise for the map. It's going to be potentially a lot of duplicate data depending on how big these vocabs can get if we're merging into a high-cardinality entity. This might be short-sighted, but part of me just wants to return the filtered mapping of study IDs to vocabularies and load that whole thing into memory and merge on the fly. This is probably a bad idea for a lot of reasons though.

**Comment by d-callan** [comment]
No, not a bad idea. I can work w that I think without too much more work than what I'd have otherwise done. And I suspect the merge service might prefer it too.

**Comment by d-callan** [comment]
I'd like @ryanrdoherty thoughts I think about the merge service...

**Comment by dmgaldi** [comment]
Yeah, I'll raise as a topic for after our scrum tomorrow. We can pull you in as well if you're free. I need to think about this a little more; it's possible automatically doing the work in subsetting/merging service wouldn't be too bad.

**Comment by bobular** [comment]
I can't comment on how useful that will be to consumer services, but aren't all the counts going to be zero?  Oh, no, maybe not, they are the counts of zeroes? Who would consume those counts?

Also, I'm curious how the study-specific vocabulary is calculated. Presumably it's the same thing as the vocabulary in the study metadata? Is it pre-calculated/stored in db at load-time?

**Comment by bobular** [comment]
I assume what the data service plugins need is, instead of a stream of data looking something like this

```
2001-02-03  SpeciesB  55
2001-02-03  SpeciesC  123
2001-02-03  SpeciesE  99
2001-02-04  SpeciesB  1
2001-02-04  SpeciesD  22
```
it would get something like this
```
2001-02-03  SpeciesA  0
2001-02-03  SpeciesB  55
2001-02-03  SpeciesC  123
2001-02-03  SpeciesD  0
2001-02-03  SpeciesE  99
2001-02-04  SpeciesA  0
2001-02-04  SpeciesB  1
2001-02-04  SpeciesC  0
2001-02-04  SpeciesD  22
2001-02-04  SpeciesE  0
```

**Comment by dmgaldi** [comment]
Thanks Bob, that makes sense. After some discussion with Danielle and Ryan, we're leaning toward an endpoint that can provide the mapping from study to vocabulary which can be used by data service to impute zeroes where necessary. This has the benefit that it should be quick to compute and won't require us to append the vocab to every row in the sample-space.

**Comment by bobular** [comment]
OK thanks for the update @dmgaldi - I leave it in your six capable hands.

Just checking that when we say "study" we mean the study entity in a megastudy, otherwise the whole study for non-megastudies (though there are no current plans to require special zero handling for these)

**Comment by d-callan** [comment]
I think probably we just mean root most entity..

**Comment by bobular** [comment]
Yes, definitely for the megastudy. That would save hardcoding any entity ids
